### PR TITLE
IGNITE-22579: Sql. Add planner test to verify numeric type coercion for IN operator

### PR DESCRIPTION
**Pull request by lowka**
Test cases for IN operator type coercion.

https://issues.apache.org/jira/browse/IGNITE-22579

---

Thank you for submitting the pull request.

To streamline the review process of the patch and ensure better code quality
we ask both an author and a reviewer to verify the following:

### The Review Checklist
- [ ] **Formal criteria:** TC status, codestyle, mandatory documentation. Also make sure to complete the following:  
\- There is a single JIRA ticket related to the pull request.  
\- The web-link to the pull request is attached to the JIRA ticket.  
\- The JIRA ticket has the Patch Available state.  
\- The description of the JIRA ticket explains WHAT was made, WHY and HOW.  
\- The pull request title is treated as the final commit message. The following pattern must be used: IGNITE-XXXX Change summary where XXXX - number of JIRA issue.
- [ ] **Design:** new code conforms with the design principles of the components it is added to.
- [ ] **Patch quality:** patch cannot be split into smaller pieces, its size must be reasonable.
- [ ] **Code quality:** code is clean and readable, necessary developer documentation is added if needed.
- [ ] **Tests code quality:** test set covers positive/negative scenarios, happy/edge cases. Tests are effective in terms of execution time and resources.

### Notes
- [Apache Ignite Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Java+Code+Style+Guide)

### CODE REVIEW

File: modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/planner/datatypes/NumericInTypeCoercionTest.java [hunk 1]
```diff
@@ -0,0 +1,1046 @@
...
+                TestBuilders.table()
+                        .name("T1")
+                        .distribution(IgniteDistributions.single())
+                        .addColumn("C1", typePair.first())
+                        .addColumn("C2", typePair.first())
+                        .addColumn("C3", typePair.second())
+                        .build()
+        );
+
+        assertPlan("SELECT c1 FROM T1 WHERE c1 IN (c2, c3)", schema, isInstanceOf(IgniteTableScan.class).and(t -> {
```
Comment (ygerzhedovich) [inline modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/planner/datatypes/NumericInTypeCoercionTest.java:1009]: Let's also add tests for literals and for dynamic parameters

File: modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/planner/datatypes/NumericInTypeCoercionTest.java [hunk 2]
```diff
@@ -0,0 +1,2412 @@
...
+import org.junit.jupiter.params.provider.MethodSource;
+
+/**
+ * A set of test to verify behavior of type coercion for IN operator, when operands belongs to the NUMERIC type family.
+ *
+ * <p>This tests aim to help to understand in which cases implicit casts are added to operands of the IN operator.
+ */
+public class NumericInTypeCoercionTest extends BaseTypeCoercionTest {
+
+    private static Stream<Arguments> nonDecimals() {
```
Comment (ygerzhedovich) [inline modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/planner/datatypes/NumericInTypeCoercionTest.java:48]: What does mean 'nonDecimal'? I see DECIMAL here

File: modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/planner/datatypes/NumericInTypeCoercionTest.java [hunk 3]
```diff
@@ -0,0 +1,3267 @@
...
+                        NumericPair.DECIMAL_3_1_DECIMAL_3_1,
+                        ofTypeWithoutCast(defaultDecimal),
+                        ofTypeWithoutCast(Types.DECIMAL_3_1),
+                        ofTypeWithoutCast(defaultDecimal),
+                        ofTypeWithoutCast(Types.DECIMAL_3_1)
+                ),
+
+                Arguments.of(
+                        NumericPair.DECIMAL_3_1_DECIMAL_3_1,
+                        ofTypeWithoutCast(defaultDecimal),
```
Comment (zstan) [inline modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/planner/datatypes/NumericInTypeCoercionTest.java:2041]: 4 copy-paste initializations near

File: modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/planner/datatypes/NumericInTypeCoercionTest.java [hunk 4]
```diff
@@ -0,0 +1,3267 @@
...
+                Arguments.of(
+                        NumericPair.DECIMAL_6_1_DECIMAL_6_1,
+                        ofTypeWithoutCast(defaultDecimal),
+                        ofTypeWithoutCast(Types.DECIMAL_6_1),
+                        ofTypeWithoutCast(defaultDecimal),
+                        ofTypeWithoutCast(Types.DECIMAL_6_1)
+                ),
+
+                Arguments.of(
+                        NumericPair.DECIMAL_6_1_DECIMAL_6_1,
```
Comment (zstan) [inline modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/planner/datatypes/NumericInTypeCoercionTest.java:2296]: copy-paste near

**Comment by lowka** [comment]
Updated `checkIncludesAllNumericTypePairs` to check for duplicates.

File: modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/planner/datatypes/NumericInTypeCoercionTest.java (hunk 4 above)
Comment (lowka) [reply modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/planner/datatypes/NumericInTypeCoercionTest.java]: Thanks. Added a method that checks for duplicate pairs.

File: modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/planner/datatypes/NumericInTypeCoercionTest.java (hunk 3 above)
Comment (lowka) [reply modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/planner/datatypes/NumericInTypeCoercionTest.java]: Fixed.

File: modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/planner/datatypes/NumericInTypeCoercionTest.java (hunk 2 above)
Comment (lowka) [reply modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/planner/datatypes/NumericInTypeCoercionTest.java]: Renamed to `lhsNonDecimal`.

**Review by zstan — APPROVED** [review]

**Review by ygerzhedovich — APPROVED** [review]

**Pull request merged**